src/Features/CSharp/Portable/Completion/CompletionProviders/SpeculativeTCompletionProvider.cs

```diff
@@ -58,12 +58,15 @@ private async Task<bool> ShouldShowSpeculativeTCompletionItemAsync(Document docu
                 return false;
             }
 
-            // If we're in a generic type argument context, use the start of the generic type name
-            // as the position for the rest of the context checks.
-            var testPosition = position;
             var leftToken = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
 
             var semanticModel = await document.GetSemanticModelForNodeAsync(leftToken.Parent, cancellationToken).ConfigureAwait(false);
+
+            // If we're in a ref type or a generic type argument context, use the start of the ref/generic type
+            // as the position for the rest of the context checks.
+            var testPosition = position;
+            var prevToken = leftToken.GetPreviousTokenIfTouchingWord(position);
```

Comment:
I admit it's really confusing that this code has both testPosition and prevToken _and_ tokenLeftOfType _and_ leftToken. (Also pray nobody uses just 'position' by accident!) Specifically, the final query seems to be using all of them that are in scope, and it's unclear why the inner loops aren't also setting prevToken or leftToken...

I'm also not sure why the code that is handling the "in a generic type argument" case isn't just chasing _only_ a token to the start, and then at the very end setting testPosition = tokenLeftOfType.SpanStart. Ditto for some other places -- it seems like once it knows it has to chase to an earlier token, just speaking in tokens and a final "grab the start position" should be easier at the end. Can it be broken into a flow more like:

```
var priorToken = syntaxTree.FindTokenOnLeftOfPosition(position).GetPreviousTokenIfTouchingWord();
var tokenStartingDeclaration = priorToken;

if (syntaxTree.IsGenericTypeArgumentContext(....))
{
    tokenStartingDeclaration = /* do the walk up with the loop and code that already exists */
}

if (tokenStartingDeclaration.GetAncestor<RefTypeSyntax>() != null)
{
    tokenStartingDeclaration = tokenStartingDeclaration.GetAncestor<RefTypeSyntax>().GetFirstToken();
}

if (tokenStartingDeclaration != priorToken)
{
    // if the token changed, then it means we found we had to go back, so let's query for that position
    position = tokenStartingDeclaraion.SpanStart;
}

// only use position from here on out and you have the same set of checks calling IsMemberDeclarationConext
```

Feel free to just say "no", but I can't help but wonder if we can clean this up. If nothing else, its' hard to tell if the lack of code updating prevToken and leftToken is a feature or a bug...

Reply:
somewhat agree with jason.  However, i would change the suggestion slightly.  I think we can just simply break this up into several helper methods that are called.  The topmost function then just does like 4-5 things, with a clear description of waht it's doing.  and the helpers are just responsible for each of those major algorithm ideas.  i.e. "this is the part where we walk up the right side if apropriate" and then we call into the helper for that.  etc. etc.

Reply:
I have the same feelings reading the code, and I like your ideas. If you're ok with refactoring, I'll try that out next week or so.

Reply:
If you want to give it a shot I'm more than happy to wait and also more than happy to review the result. My biggest worry asking contributors to clean up stuff like this is not having to deal with the result, but rather there's always a certain amount of "hey, you found a mess we made, want to clean it up for us?" and that makes me feel guilty.

But if it's really too icky I'm OK with taking this. Even if the code is tricky to decipher, like I said your tests are great so I at least have confidence it works!